### PR TITLE
Feed api implementation

### PR DESCRIPTION
**Pull request by Jacquigee**
# Scope

_Please make sure to read the [Contribution Guidelines](https://github.com/droidconKE/droidconKE2023Android/blob/main/CONTRIBUTING.md)
and check that you understand and have followed it as best as possible Explain what your feature
does in a short paragraph._ please check the below boxes
- [X] I have followed the coding conventions
- [X] I have added/updated necessary tests
- [X] I have tested the changes added on a physical device
- [X] I have run `./codeAnalysis.sh` to make sure all lint/formatting checks have been done.

## Closes/Fixes Issues
_Declare any issues by typing `fixes #1` or `closes #1` for example so that the automation can kick
in when this is merged_

## Other testing QA Notes
_What have you tested specifically and what possible impacts/areas there are that may need retesting
by others._

Please add a screenshot (if necessary)

The current response from the server.

Feed_Response

### CODE REVIEW

File: data/src/main/java/com/android254/data/repos/FeedManager.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                ResourceResult.Success(
+                    data.map { it.toDomain() }
+                )
+            }
+
+            else -> {
+                ResourceResult.Success(emptyList())
+            }
+        }
+    }
```
Comment (chepsi) [inline data/src/main/java/com/android254/data/repos/FeedManager.kt:52]: here you should make the call from `FeedApi` -> `feedApi. fetchFeed()`

File: domain/src/main/java/com/android254/domain/repos/FeedRepo.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android254.domain.repos
+
+import com.android254.domain.models.Feed
+import com.android254.domain.models.ResourceResult
+
+interface FeedRepo {
+    suspend fun fetchFeed(): ResourceResult<List<Feed>>
```
Comment (chepsi) [inline domain/src/main/java/com/android254/domain/repos/FeedRepo.kt:22]: `ResourceResult<List<Feed>>` -> `List<Feed>`?

File: presentation/src/main/java/com/android254/presentation/common/bottomsheet/Strings.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    val resources = LocalContext.current.resources
+    return when (string) {
+        Strings.NavigationMenu -> resources.getString(R.string.navigation_menu)
+        Strings.CloseDrawer -> resources.getString(R.string.close_drawer)
+        Strings.CloseSheet -> resources.getString(R.string.close_sheet)
+        Strings.DefaultErrorMessage -> resources.getString(R.string.default_error_message)
+        Strings.ExposedDropdownMenu -> resources.getString(R.string.dropdown_menu)
+        Strings.SliderRangeStart -> resources.getString(R.string.range_start)
+        Strings.SliderRangeEnd -> resources.getString(R.string.range_end)
+        else -> ""
```
Comment (chepsi) [inline presentation/src/main/java/com/android254/presentation/common/bottomsheet/Strings.kt:50]: Can't we reference this directly from the composable?

File: presentation/src/main/java/com/android254/presentation/feed/FeedViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class FeedViewModel @Inject constructor(
+    private val feedRepo: FeedRepo
+) : ViewModel() {
+    private val _feedsState = MutableStateFlow<FeedUIState>(FeedUIState.Loading)
+    val feedsState: StateFlow<FeedUIState> get() = _feedsState
```
Comment (chepsi) [inline presentation/src/main/java/com/android254/presentation/feed/FeedViewModel.kt:35]:
```
var viewState by mutableStateOf(FeedUIState)
private set
```

File: presentation/src/main/java/com/android254/presentation/feed/view/FeedComponent.kt [hunk 5]
```diff
@@ -135,7 +136,11 @@ fun FeedComponent(modifier: Modifier, onClickItem: (Int) -> Unit) {
 @Composable
 fun Preview() {
     ChaiDCKE22Theme {
-        FeedComponent(modifier = Modifier) {
-        }
+        FeedComponent(
+            modifier = Modifier,
+            feedPresentationModel =
+            FeedUI("Feed", "Feed feed", "test", "", "", ""),
```
Comment (chepsi) [inline presentation/src/main/java/com/android254/presentation/feed/view/FeedComponent.kt:142]: minor: formatting?

**Review by chepsi — COMMENTED** [review]
Left some comments

File: presentation/src/main/java/com/android254/presentation/feed/view/FeedComponent.kt (hunk 5 above)
Comment (Jacquigee) [reply presentation/src/main/java/com/android254/presentation/feed/view/FeedComponent.kt:142]: What do you mean by this?

Comment (chepsi) [reply presentation/src/main/java/com/android254/presentation/feed/view/FeedComponent.kt:142]: lint formatting of the file.

File: data/src/main/java/com/android254/data/repos/FeedManager.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    private val FeedApi: FeedApi
+) : FeedRepo {
+    override suspend fun fetchFeed(): List<Feed>? {
+        return when (val result = FeedApi.fetchFeed(1, 100)) {
+            is DataResult.Empty -> {
+                emptyList()
+            }
+
+            is DataResult.Error -> {
+                null
```
Comment (wangerekaharun) [inline data/src/main/java/com/android254/data/repos/FeedManager.kt:35]: We should handle this case

File: presentation/src/main/java/com/android254/presentation/common/bottomsheet/BottomSheetScaffold.kt
Comment (wangerekaharun) [inline presentation/src/main/java/com/android254/presentation/common/bottomsheet/BottomSheetScaffold.kt:1]: @Jacquigee these files were removed a while back, is there a reason we are adding them back?

File: presentation/src/main/java/com/android254/presentation/feed/FeedViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+class FeedViewModel @Inject constructor(
+    private val feedRepo: FeedRepo
+) : ViewModel() {
+    var viewState: FeedUIState by mutableStateOf(FeedUIState.Loading)
+        private set
+
+    fun fetchFeed() {
+        viewModelScope.launch {
+            viewState = when (val value = feedRepo.fetchFeed()) {
+                null -> FeedUIState.Error("Error getting result")
```
Comment (wangerekaharun) [inline presentation/src/main/java/com/android254/presentation/feed/FeedViewModel.kt:40]: change these as per the comment on the  repo file

File: presentation/src/main/java/com/android254/presentation/feed/view/FeedScreen.kt [hunk 8]
```diff
@@ -16,76 +16,86 @@
...
-    )
-
-    BackHandler(bottomSheetState.isVisible) {
-        scope.launch { bottomSheetState.hide() }
-    }
-
-    Scaffold(
+    feedViewModel.fetchFeed()
+    val feedUIState = feedViewModel.viewState
+    BottomSheetScaffold(
```
Comment (wangerekaharun) [inline presentation/src/main/java/com/android254/presentation/feed/view/FeedScreen.kt:53]: Material 3 has Modal Bottom sheet, we can use that instead

File: presentation/src/main/java/com/android254/presentation/feed/view/FeedScreen.kt [hunk 9]
```diff
@@ -16,76 +16,86 @@
...
-            LazyColumn(
-                modifier = Modifier.testTag("feeds_lazy_column"),
-                verticalArrangement = Arrangement.spacedBy(4.dp)
-            ) {
-                items(count = 10) {
-                    FeedComponent(modifier = Modifier.fillMaxWidth()) {
-                        scope.launch {
-                            bottomSheetState.show()
+            when (feedUIState) {
+                is FeedUIState.Error -> {}
```
Comment (wangerekaharun) [inline presentation/src/main/java/com/android254/presentation/feed/view/FeedScreen.kt:76]: Let's handle these cases also

**Comment by wangerekaharun** [comment]
@Jacquigee Left some comments, let me know if you have questions

**Comment by Jacquigee** [comment]
Please review this PR and let me know if there are changes that I need to make.

**Review by chepsi — APPROVED** [review]

**Review by chepsi — APPROVED** [review]

**Review by wangerekaharun — APPROVED** [review]

**Pull request merged**